Code restructure + Unit Tests + Abstract class PR

qwofford opened the pull request:
This pull request tracks several issues which should all be merged at the same time: #14, #16, and #12 so far.

qwofford commented:
I created an abstract class and implementation for Plugins. I created a generic environment plugin driver based on the generic plugin driver which always captures some basic information about the process owner, and then I created a specific hostname plugin driver of the environment plugin driver type which just grabs the name of the machine and stores it in an output collector. An example of how to use it is below:
```
>>> from dsi.plugins.env import HostnamePlugin
>>> a=HostnamePlugin()
>>> a.parse()
>>> a.add_row()
>>> a.add_row()
>>> a.output_collector
OrderedDict([('uid', [1000, 1000, 1000]), ('effective_gid', [1000, 1000, 1000]), ('moniker', ['qwofford', 'qwofford', 'qwofford']), ('gid_list', [[1000, 24, 25, 29, 30, 33, 44, 46, 108, 111, 112, 117, 999, 123, 1004, 39817, 23127, 36153, 38144, 34143, 32942, 30934, 38107, 29808, 29492, 26772, 34148, 5404, 33519, 17872, 39067, 4773, 30350, 32280, 36790, 30125, 37830, 19038, 28338, 32270, 39870, 26346], [1000, 24, 25, 29, 30, 33, 44, 46, 108, 111, 112, 117, 999, 123, 1004, 39817, 23127, 36153, 38144, 34143, 32942, 30934, 38107, 29808, 29492, 26772, 34148, 5404, 33519, 17872, 39067, 4773, 30350, 32280, 36790, 30125, 37830, 19038, 28338, 32270, 39870, 26346], [1000, 24, 25, 29, 30, 33, 44, 46, 108, 111, 112, 117, 999, 123, 1004, 39817, 23127, 36153, 38144, 34143, 32942, 30934, 38107, 29808, 29492, 26772, 34148, 5404, 33519, 17872, 39067, 4773, 30350, 32280, 36790, 30125, 37830, 19038, 28338, 32270, 39870, 26346]]), ('hostname', ['MYMACHINE', 'MYMACHINE', 'MYMACHINE'])])
```

I created a few example unit tests for this hostname environment plugin driver, with output that looks like this:
```
pytest -v
=============================================================== test session starts ================================================================
platform linux -- Python 3.11.3, pytest-7.3.1, pluggy-1.0.0 -- ~/miniconda3/envs/dsi/bin/python
cachedir: .pytest_cache
rootdir: <PROJECT ROOT>/dsi/plugins/tests
collected 2 items                                                                                                                                  

test_env.py::test_hostnameplugin_type PASSED                                                                                                 [ 50%]
test_env.py::test_hostname_plugin_col_shape PASSED                                                                                           [100%]

================================================================ 2 passed in 0.01s =================================================================
```

I invite all reviewers to look at design patterns in `./dsi/plugins/*` and offer comments, or better yet, PRs into branch `qwofford/issue14` along with an explanation for the changes. This should be a good way for us to learn from each other.

@DanielRJohnson I think the provenance plugin driver could be of the Environment Plugin type, and it can go directly in env.py. If my plugin classes don't fit your specific implementation (they probably won't)...let's fix the classes I stubbed out. These are pretty rough!

qwofford commented:
I think this PR is ready to merge. 

* All existing code has been moved into the plugin/driver model. 
* All examples have been converted to unit tests. Note that you'll need to `python -m pip install gitpython` to run these Unit tests.
* All code which is no longer in use has been removed.

My `pytest -v` results below:
```
(dsi) ➜  dsi git:(qwofford/issue14) ✗ pytest -v                
================================================ test session starts =================================================
platform linux -- Python 3.11.3, pytest-7.3.1, pluggy-1.0.0 -- ~/miniconda3/envs/dsi/bin/python
cachedir: .pytest_cache
rootdir: <PROJECT ROOT>
collected 7 items                                                                                                    

dsi/drivers/filesystem/tests/test_sqlite.py::test_wildfire_data_sql_artifact PASSED                            [ 14%]
dsi/drivers/filesystem/tests/test_sqlite.py::test_wildfire_data_csv_artifact PASSED                            [ 28%]
dsi/drivers/filesystem/tests/test_sqlite.py::test_yosemite_data_csv_artifact PASSED                            [ 42%]
dsi/drivers/filesystem/tests/test_sqlite.py::test_query PASSED                                                 [ 57%]
dsi/drivers/filesystem/tests/test_sqlite.py::test_artifact_query PASSED                                        [ 71%]
dsi/plugins/tests/test_env.py::test_hostname_plugin_type PASSED                                                [ 85%]
dsi/plugins/tests/test_env.py::test_hostname_plugin_col_shape PASSED                                           [100%]

================================================= 7 passed in 0.30s ==================================================

```

qwofford commented:
I converted this back to a draft, still need some GUFI driver work. 

@hugegreenbug could you take a look at what I've done with Jesus' sqlite driver and do something similar for the GUFI back end, including fixing up the documentation table of contents in `./docs/`?

qwofford commented:
@jpulidojr Note that I removed util.py because I only found one instance of a function defined there being used, `check_type`. This function seemed to be associated with the SqlStore class, so I added it as a class method and deleted the rest of util.py. It's possible you want to save some more of this. Hugh and I were thinking some of util.py might be useful in the DSI core, but since there weren't tests in `examples`, I just removed it.

Please feel free to add utils back. I'm a little concerned about adding a catch-all directory for utility functions, so if we can make these global helpers class methods instead, that might be good!

qwofford commented:
Also @jpulidojr please look at SqlStore unit tests and see if I captured the `./examples` directory sufficiently.

qwofford commented:
Daniel and I did some work to refine the DSI Plugin concept over the last week. We have merged our changes into this PR, and this PR is again ready for review by the larger DSI developer group.

qwofford commented:
Specifically, we would like feedback on the Plugin documentation, project re-structure, and unit tests.

Converted back to a draft. We still need to perform a similar set of tasks for Drivers before merging in.

jpulidojr commented:
I went over the tests and I think for the wildfire data it is a unique enough query to be sufficient for sql testing. Just a note, I did a clean install to test this and pytest and we may need to upgrade the set of requirements, for example, someone needs to

pip3 install pytest
pip3 install gitpython

As far as the previous version, we had no dependencies outside of what was already provided by a default python install

qwofford commented:
The Bueno data I had to parse was very simple, and the implementation is complete. I rolled the Bueno plugin into this PR as it shouldn't affect anyone else.

qwofford commented:
Added git commit awareness to Plugin ABC, I think we should consider having the drivers follow the same pattern.

I added an abstract class variable which is required to implement. The first implementation of this abstract class should capture the git commit hash at pip install time. This is compatible with installation from source, and from pypa package distribution. Please read #30 for details.

qwofford commented:
I added a Parquet driver because it was a bit simpler to think through the Core Terminal Implementation with that.

qwofford commented:
Trying again...I _think_ this PR is ready to go into main. It might take some time to test. I recommend testing after reading the Core Terminal documentation: https://lanl.github.io/dsi/core.html

This explains one way we might demonstrate work with DSI to satisfy upcoming milestones...

DanielRJohnson reviewed (commented):
Here's some thoughts. Everything is minor, looks good.

qwofford commented on docs/core.rst:
```diff
@@ -0,0 +1,86 @@
+>>> a.load_module('plugin','Hostname','producer')
+>>> # Hostname plugin producer loaded successfully.
+>>> a.list_loaded_modules()
+>>> # {'producer': [<dsi.plugins.env.Hostname object at 0x7f21232474d0>], 
```
I think I like the full path to the plugin Class because it tells us that the producer functionality is implemented as a Plugin. To a Core Terminal user, this could be a good reminder that the thing they loaded is a plugin vs a driver. It also gives us the ability to throw errors if some non-sense is requested, like a plugin for a front-end.

qwofford commented on docs/core.rst:
```diff
@@ -0,0 +1,86 @@
+
+Although this demonstration only used one Plugin per Plugin functionality, any number of plugins can be added to collect an arbitrary amount of queriable metadata::
+
+>>> a.load_module('driver','Parquet','front-end',filename='parquet.data')
```
fixed

qwofford commented on dsi/drivers/tests/test_gufi.py:
```diff
@@ -0,0 +1,25 @@
+    store.close()
+    print("Done")
+
+test_artifact_query()
```
I revised this test to check for somethign meaningful

qwofford commented on dsi/drivers/tests/test_sqlite.py:
```diff
@@ -0,0 +1,55 @@
+    store = Sqlite(dbpath)
+    anames = store.get_artifact_list(isVerbose=isVerbose)
+    for name in anames:
+        print( name[0] )
```
I struggled with this one. All of these sqlite tests are about saving a file out to disk. In order to test it you'd need to inspect the file. However if the file writing failed, which is what we're testing, we would get an error by the time we hit the end of the pytest function. This test could be better, but I just added an `assert True` at the end for now and removed print statements.

qwofford commented on dsi/drivers/tests/test_gufi.py:
```diff
@@ -0,0 +1,25 @@
+    store = Gufi(prefix, index, dbpath, table, column, isVerbose)
+    sqlstr = "select * from dsi_entries"
+    rows = store.get_artifacts(sqlstr)
+    print( rows )
```
Fixed it.

qwofford commented:
Did you mean to set `qstr` on line 83 but then not use it? I changed it to an underscore if so. Underscores are the only style compliant variable name that pylama won't complain about.
